bootstrap/app/addon-manager.go

```diff
@@ -0,0 +1,306 @@
+					filterLog(&out, "", klog.Info)
+					filterLog(&errOut, "", klog.Warning)
+					if err == nil {
+						klog.Infof("== Successfully started %s in namespace default ==", path)
+						return nil
+					}
+					klog.Warningf("== Failed to start %s in namespace default. %d tries remaining. ==", path, attemptsRemaining-1)
+					if attemptsRemaining > 1 {
+						time.Sleep(startAddonRetryDelay)
+					}
+				}
+			}
+			return err
+		})
+}
+
+// IsLeader returns whether this instance of addonManager should act as leader in a multi-master cluster.
+func (m *addonManager) isLeader() bool {
+	var out, errOut bytes.Buffer
+	klog.Info("== Determining addon-manager leader ==")
+	// TODO: Do a real leader election instead of piggybacking off of KCM
+	err := m.kubectl(&out, &errOut, m.kubectlOpts, "-n", systemNamespace, "get", "ep", "kube-controller-manager", "-o", `go-template={{index .metadata.annotations "control-plane.alpha.kubernetes.io/leader"}}`)
+	filterLog(&errOut, "", klog.Warning)
+	if err != nil {
+		// If the leader can't be positively established, assume that we're it
+		return true
+	}
+	holderIdentity := extractHolderIdentity(out.String())
+	return holderIdentity == "" || strings.HasPrefix(holderIdentity, m.hostname+"_")
+}
+
+func extractHolderIdentity(raw string) string {
+	captureGroups := leaderRegexp.FindStringSubmatch(raw)
+	if len(captureGroups) != 2 {
+		// A match yields one capture group (index 1)
+		return ""
+	}
+	return captureGroups[1]
+}
+
+func (m *addonManager) reconcileAddons() {
+	var out, errOut bytes.Buffer
+	klog.Info("== Reconciling with deprecated label ==")
+
+	// TODO: Remove the first command in future release.
+	// Adding this for backward compatibility. Old addons have CLUSTER_SERVICE_LABEL=true and don't have
+	// ADDON_MANAGER_LABEL=EnsureExists will still be reconciled.
+	m.kubectl(&out, &errOut, append([]string{
+		m.kubectlOpts,
+		"apply",
+		"-f", m.addonPath,
+		"--recursive",
+		"-l", clusterServiceLabel + "=true," + addonManagerLabel + "!=" + modeEnsureExists,
+		"--prune=true"},
+		m.pruneWhitelistFlags...)...)
+	// Filter out `configured` message to not noisily klog. `created`, `pruned` and errors will be logged.
+	filterLog(&out, "configured", klog.Info)
+	filterLog(&errOut, "configured", klog.Warning)
+	out.Reset()
+	errOut.Reset()
+	klog.Info("== Reconciling with addon-manager label ==")
+
+	m.kubectl(&out, &errOut, append([]string{
+		m.kubectlOpts,
+		"apply",
+		"-f", m.addonPath,
+		"--recursive",
+		"-l", clusterServiceLabel + "!=true," + addonManagerLabel + "=" + modeReconcile,
+		"--prune=true"},
+		m.pruneWhitelistFlags...)...)
+	filterLog(&out, "configured", klog.Info)
+	filterLog(&errOut, "configured", klog.Warning)
+	klog.Info("== Kubernetes addon reconcile completed ==")
+}
+
+func (m *addonManager) ensureAddons() {
+	var out bytes.Buffer
+	m.kubectl(&out, &out,
+		m.kubectlOpts,
+		"create",
+		"-f", m.addonPath,
+		"--recursive",
+		"-l", addonManagerLabel+"="+modeEnsureExists)
```

Comment:
Meta: I find the various options here confusing (e.g. that we're not pruning here).  But I realize you're translating existing behaviour!

Reply:
I'm with ya :p